### ISSUE TITLE
filter() does not seem to work when passed through middleware

### ISSUE THREAD

**Issue by swhite2** [onury/accesscontrol#77]
Hi there, loving your work! I am however running into an issue with the `permission.filter()` function. It seems ES6's native implementation of filter() is called when a `Permission` object is passed through the app with `res.locals.permission`. Which means that if I try to do this: `res.json(res.locals.permission.filter(req.user));` it returns an empty object.

My models are defined as `.readOwn('user', [ 'name', 'email', 'phone' ])`, and according to the documentation I should expect just the name, email and phone as a response.

**Comment by AndreasWJ** [comment]
Found any fix or workaround?

Have the same problem, however I'm just passing the filter function forward by binding the permission object.
`req.filterFn = permission.filter.bind(permission);`

**Comment by TheLartians** [comment]
I've run into a similar issue using Mongoose Documents, where `permission.filter(user)` returned empty objects. My current workaround is converting the document to an object first via `permission.filter(user.toObject())`.

**Comment by noclat** [comment]
Yep, that's because `res.locals` serializes data, it's not meant to hold instances. You can pass them across middlewares using the `req` object, like `req.permissions = [/* ... */]` and there you'd be able to access all its methods.
Though, AccessControl exposes the `filter` method on its instance as well (`ac.filter()`), so maybe there's no need to access it from the permission.

**Comment by onury** [comment]
In v3, `permission.filter(data)` returns a shallow copy containing only the granted attributes — it works on a plain object or an array of objects. An empty result almost always means one of:

- **The check wasn't granted**, so there are no attributes to keep — check `permission.granted` first. For `readOwn`, v3 only grants when ownership is satisfied: configure `policy.ownerField`/`owner` and pass the record in context (or use `readAny` if any-record read is intended).
- **You're calling `.filter()` on something that isn't the live `Permission`** — e.g. a value that was JSON-serialized and revived, which drops the class methods (so JS's native `Array.prototype.filter` gets involved). Keep the `Permission` instance and call `.filter()` on it directly, before serializing.

```js
const perm = ac.can(role).readOwn('user');           // live Permission
if (perm.granted) res.json(perm.filter(req.user));   // { name, email, phone }
```

See the [filtering recipe](https://onury.io/accesscontrol/guides/recipes/). Closing, as this is usage on the (pre-v3) API — please open a fresh issue with a small v3 repro if you still hit it.

Sorry for the late reply!
